REFACTOR-#5079: Add mypy checks for modin.core.dataframe.base

shalearkane opened the pull request:
## What do these changes do?
- added type hints to functions in modin.core.dataframe.base
- also added "# type: ignore" to abstract methods to ignore non-existent class members

## Tasks
- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s`
- [x] Resolves #5079 
- [x] tests passing `pytest modin/test/test_*` with output `140 passed, 98 warnings`

Signed-off-by: Soumik Dutta <shalearkane@gmail.com>

mvashishtha commented on modin/core/dataframe/base/partitioning/axis_partition.py:
```diff
@@ -95,9 +96,11 @@ def _wrap_partitions(self, partitions):
             List of wrapped remote partition objects.
         """
-        return [self.partition_type(obj) for obj in partitions]
+        return [self.partition_type(obj) for obj in partitions]  # type: ignore
```
Why do we need the ignore?

mvashishtha replied:
@shalearkane thank you, that makes sense. I suggest that we assert within this function that `instance_type` and `partition_type` are not `None`, because subclasses of `BaseDataframeAxisPartition` should set those fields to a non-`None` value.

mvashishtha commented on modin/core/dataframe/base/partitioning/axis_partition.py:
```diff
@@ -114,9 +117,9 @@ def force_materialization(self, get_ip=False):
             lambda x: x, num_splits=1, maintain_partitioning=False
         )
-        return type(self)(materialized, get_ip=get_ip)
+        return type(self)(materialized, get_ip=get_ip)  # type: ignore[call-arg]
```
Why do we need the ignore?

shalearkane replied:
Because we have not yet defined an `__init__` function in class `BaseDataframeAxisPartition` with the args `materialized` and `get_ip`, mypy is outputting the following messages:
```
modin/core/dataframe/base/partitioning/axis_partition.py:120:16: error: Too many arguments for "BaseDataframeAxisPartition"  [call-arg]
modin/core/dataframe/base/partitioning/axis_partition.py:120:16: error: Unexpected keyword argument "get_ip" for "BaseDataframeAxisPartition"  [call-arg]
```

If we implement the following method, mypy doesn't emit any warning
```python
def __init__(self, *args: Iterable, **kwargs: bool) -> None:
        super().__init__()
```

mvashishtha commented on modin/core/dataframe/base/partitioning/axis_partition.py:
```diff
@@ -114,9 +117,9 @@ def force_materialization(self, get_ip=False):
+        return type(self)(materialized, get_ip=get_ip)  # type: ignore[call-arg]
 
-    def unwrap(self, squeeze=False, get_ip=False):
+    def unwrap(self, squeeze: bool = False, get_ip: bool = False) -> Union[list,Tuple[Any, Any]]:
```
Why do we need the ignores here?

mvashishtha reviewed (commented):
@shalearkane thank you very much for making this PR. I have left some inline comments.

Could you please also add `modin/core/dataframes/base/` to the `mypy.ini` file? After that, you can run `mypy --config-file mypy.ini modin/core/dataframes/base/` locally to test your changes. Please leave comments here if you have any questions or you get stuck.

shalearkane commented on modin/core/dataframe/base/partitioning/axis_partition.py:
```diff
@@ -138,13 +141,13 @@ def unwrap(self, squeeze=False, get_ip=False):
         (i.e. [(Ray.ObjectRef/Dask.Future, Ray.ObjectRef/Dask.Future), ...]).
         """
-        if squeeze and len(self.list_of_blocks) == 1:
+        if squeeze and len(self.list_of_blocks) == 1:  # type: ignore[attr-defined]
```
Yes, we could define them in the base class. So what should be the type of `list_of_blocks` and `list_of_ips` ... should I define the type as `Optional[list]` or just `list`?

pyrito replied:
@shalearkane if it's possible for the list to be `None` then it should be `Optional[list]` otherwise we should define the list as an empty list.

mvashishtha replied:
Neither list should be none. @shalearkane could you try adding `@property` `@abstractmethod` methods specifying a non-null `list` for both lists? Once we have those, I think mypy should let us assume we have lists.

For reference:

- Only implementation for `list_of_blocks` is [in axis_partition](https://github.com/modin-project/modin/blob/d6d503ac7c3028d871c34d9e99e925ddb0746df6/modin/core/dataframe/pandas/partitioning/axis_partition.py#L32)
- Only implementations for `list_of_ips` are [ray](https://github.com/modin-project/modin/blob/238a42875ce585de63932c8beefccb31f0d28cd6/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/virtual_partition.py#L134) and [dask](https://github.com/modin-project/modin/blob/d6d503ac7c3028d871c34d9e99e925ddb0746df6/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/virtual_partition.py#L132) virtual partitions.

mvashishtha reviewed (commented):
@shalearkane I have responses to some open threads, but not all. I'll get to the rest soon.

Also, **could you please also add modin/core/dataframes/base/ to the mypy.ini file?** After that, you can run mypy --config-file mypy.ini modin/core/dataframes/base/ locally to test your changes.

mvashishtha reviewed (commented):
@shalearkane I have left some comments and more suggestions. I resolved one thread. Please let me know if you need any more help. I would really like to help you merge this PR.

shalearkane commented on mypy.ini:
```diff
@@ -18,7 +18,8 @@ warn_unreachable=True
         modin/logging/*.py,
-        modin/*.py
+        modin/*.py,
+        modin/core/dataframe/base/**/*.py
```
Okay, I will resolve the issues soon.

pyrito commented:
Hi @shalearkane what's the status of this PR? We would love to help you get this merged into Modin if possible 😃 🎉

shalearkane commented:
Yes I am working on the PR, I will update the PR by 10PM IST.

mvashishtha commented:
@shalearkane I would like to request the following:

1) remove the `list_of_ips` definition and go back to marking all the `list_of_ips` calls with `  # type: ignore[attr-defined]`. 
2) Above the new attr-defined ignores, please add a single comment like  `TODO(https://github.com/modin-project/modin/issues/5176): Stop ignoring the list_of_ips check once we know that we're not calling list_of_ips on python axis partitions`.
3) Run mypy locally to check that it's passing.
4) push your commits

shalearkane commented:
Yes, I removed `list_of_ips` however mypy was giving these errors  in `modin/logging/config.py` :
```
modin/logging/config.py:168:17: error: Call to untyped function "virtual_memory" in typed context  [no-untyped-call]
modin/logging/config.py:195:17: error: Call to untyped function "virtual_memory" in typed context  [no-untyped-call]
``` 
So I added `type: ignore` to those files

mvashishtha reviewed (commented):
@shalearkane I'm leaving one inline comment. There are also a few CI failures that you can see at the bottom of the conversation page. Before you address them, please run `pip install -r requirements-dev.txt` from your modin source directory. to make sure your local testing tools are up to date.

- [black](https://github.com/modin-project/modin/actions/runs/3380747973/jobs/5648016294): Please run `black modin/ asv_bench/benchmarks scripts/doc_checker.py` to fix these, then `black --check --diff modin/ asv_bench/benchmarks scripts/doc_checker.py` to check that black passes.
- [flake8](https://github.comproject/modin/actions/runs/3380747973/jobs/5648016634): Please address all these failures and check your work with `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [mypy](https://github.com/modin-project/modin/actions/runs/3380747973/jobs/5648016528): mypy is actually complaining that the ignores in modin/logging don't do anything. Which version of mypy are you using? GitHub CI uses mypy 0.982, which doesn't seem to have a problem with the `virtual_memory` call.

shalearkane commented:
I was using mypy version ```0.942``` the default mypy which comes preinstalled with Ubuntu 22.04. But yes, running mypy ```0.942``` does not produce errors.

mvashishtha reviewed (approved):
@shalearkane LGTM, thank you!

mvashishtha commented:
@shalearkane I'm going to merge this PR now. Thank you very much for contributing this PR. It was not simple and you had to respond to many comments. Please continue to reach out on GitHub or other Modin forums to contribute to Modin.
